doc/source/whatsnew/v1.3.0.rst

```diff
@@ -753,6 +753,7 @@ Groupby/resample/rolling
 - Bug in :class:`core.window.ewm.ExponentialMovingWindow` when calling ``__getitem__`` would not retain ``com``, ``span``, ``alpha`` or ``halflife`` attributes  (:issue:`40164`)
 - :class:`core.window.ewm.ExponentialMovingWindow` now raises a ``NotImplementedError`` when specifying ``times`` with ``adjust=False`` due to an incorrect calculation (:issue:`40098`)
 - Bug in :meth:`Series.asfreq` and :meth:`DataFrame.asfreq` dropping rows when the index is not sorted (:issue:`39805`)
+- Bug in :meth:`DataFrameGroupBy.aggregate`, :meth:`SeriesGroupBy.aggregate`, :meth:`DataFrameGroupBy.transform`, and :meth:`SeriesGroupBy.transform` would possibly change the result dtype when ``func`` is callable (:issue:`21240`)
```

Comment:
i think this needs a sub-section it can be an expected change (agree with the change though)

Reply:
may also need to update groupby.rst, we *might* have this documented

Reply:
Can do - will have a look through the guide and make updates as appropriate. Still have concerns with some of the behavior changes this PR is making as noted in the OP (and results in some failed tests that I can fixup). Namely, if the input dtype is e.g. float32 and we go through the fastpath, output in this PR (assuming it's still a float) will be float64 instead of float32. On master, we'd get back float32. Not sure if there is a good way to solve this.

Reply:
Ah - ignore the above, I should have looked closer at the implementation of maybe_downcast_numeric.